fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/MatchLeafRule.java

```diff
@@ -0,0 +1,126 @@
+import com.apple.foundationdb.record.query.plan.temp.IdentityBiMap;
+import com.apple.foundationdb.record.query.plan.temp.IterableHelpers;
+import com.apple.foundationdb.record.query.plan.temp.MatchCandidate;
+import com.apple.foundationdb.record.query.plan.temp.MatchInfo;
+import com.apple.foundationdb.record.query.plan.temp.PartialMatch;
+import com.apple.foundationdb.record.query.plan.temp.PlanContext;
+import com.apple.foundationdb.record.query.plan.temp.PlannerRule;
+import com.apple.foundationdb.record.query.plan.temp.PlannerRuleCall;
+import com.apple.foundationdb.record.query.plan.temp.RelationalExpression;
+import com.apple.foundationdb.record.query.plan.temp.matchers.ExpressionMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.ListChildrenMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.TypeMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matching.BoundMatch;
+
+import javax.annotation.Nonnull;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Expression-based transformation rule that matches any leaf expression. The rule creates matches of type
+ * {@link PartialMatch} for any match between this expression and a leaf expression in a {@link MatchCandidate}.
+ * It seeds the memoization structure for partial matches that is kept as part of {@link ExpressionRef}. It prepares
+ * further rules such as {@link MatchIntermediateRule} and {@link AdjustMatchRule}.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class MatchLeafRule extends PlannerRule<RelationalExpression> {
+    // match any relational expression that is a leaf, that is, any expression that does not have any children
+    private static final ExpressionMatcher<RelationalExpression> root =
+            TypeMatcher.of(RelationalExpression.class, ListChildrenMatcher.empty());
+
+    public MatchLeafRule() {
+        super(root);
+    }
+
+    /**
+     * Note: Transformation rules for expressions are partitioned by the class of the root matcher. This does not work
+     * here as this rule is non-specific which means that it matches sub classes of {@link RelationalExpression} and not
+     * just {@link RelationalExpression} itself. In order for this rule to fall into the set of rules that is always
+     * utilized we return {@code Optional.empty()} here.
+     * @return {@code Optional.empty()}
+     */
+    @Override
+    public Optional<Class<? extends Bindable>> getRootOperator() {
+        return Optional.empty();
+    }
+
+    @Override
+    public void onMatch(@Nonnull PlannerRuleCall call) {
+        final PlanContext context = call.getContext();
+        final RelationalExpression expression = call.get(root);
+        // iterate through all candidates known to the context
+        for (final MatchCandidate matchCandidate : context.getMatchCandidates()) {
+            final ExpressionRefTraversal traversal = matchCandidate.getTraversal();
```

Comment:
This was discussed in the review meeting also (because of this comment). The main point why I believe this is okay is that the matching rules only traverse the candidate graph, never the query query graph itself. The candidate graphs at this point are maximally pruned at that point, so in a sense references are meaningless on the candidate side and can be skipped and not exposed.

Reply:
Opened #1124 to track this.